### PR TITLE
Voucher bare minimum backoffice

### LINKED ISSUE #10431: [Vouchers] Implement bare minimum backoffice
## Description

The bare minimum backend implements the initial logic to create a voucher. These vouchers will be vastly simplified from the [final designs](https://www.figma.com/file/ZOe1Xmz4kg9wnHw1XFu389/Voucher-Back-office?node-id=47%3A301)

Two backoffice pages in Enterprise Settings:
1. Add Voucher Page
User can include a Voucher code - typed
No option to 'generate code'
Voucher type limited to 'flat amount'
Amount limited to $10
Expires and date not included - vouchers are infinite by default
Leave off 'voucher purpose' (later we can default to food equity)
Leave off 'Label'

2. List Vouchers Page
Leave off all filters
Include 'add new' button
Include table with cols 'Voucher Code', 'Rate'. All else dropped.
No delete yet.

At time of writing there was still some debate/exploration of how the data model will work - will we use the existing calculators and adjustments functionality? Will this fit in with tax implications? The plan is make as few implementation decisions as possible about this while we check in with AU accountant. Please document any assumptions that are made along the way.

## Acceptance Criteria & Tests

1. Simplified BE implementation as described above
2. We won't have a checkout version to test this yet, so we can just test the creation of Vouchers are created in data model and thus  list page as expected
3. Translations as standard
4. Unit tests as standard

### PR DESCRIPTION
**Pull request by rioug**
#### What? Why?

- Closes #10431 

Implement a bare minimum admin screen for vouchers. It allows you to create a voucher with a flat value of $10. Currently you can only enter one code at the time.

#### What should we test?

Enable vouchers feature 
- Visit `admin/feature-toggle/features` and enable or add `vouchers` feature

In the backoffice logged as an enterprise user:
- Visit the Enterprise tab and click on "Settings" for a shop of your choosing
- Check that 'vouchers' appears in the left menu column, and click on it 
  --> It should load a page saying "No vouchers yet" and have "Add New" button
- Click on "Add New button"
  --> It should load the New Voucher page 
- Enter a voucher code and click "Save"
 --> It should redirect you to the vouchers page on the enterprise edit page
 --> You should now see a list of vouchers with the one you just created
- Go back to the New Voucher page
- Click on Save without entering any code
  --> You should see an error message saying "Code can't be blank"  

#### Release notes

Changelog Category: User facing changes

The title of the pull request will be included in the release notes.

### CODE REVIEW

**Comment by rioug** [comment]
Do I need to add some script to create the "vouchers" feature toggle ?

**Comment by mkllnk** [comment]
> Do I need to add some script to create the "vouchers" feature toggle ?

Just use it as you do. We don't need to restrict the routes. I think that a toggle on UI level is enough.

File: spec/models/voucher_spec.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    it { is_expected.to belong_to(:enterprise) }
+  end
+
+  describe 'validations' do
+    subject { Voucher.new(code: 'new_code', enterprise: enterprise) }
+
+    let(:enterprise) { build(:enterprise) }
+
+    it { is_expected.to validate_presence_of(:code) }
+    it { is_expected.to validate_uniqueness_of(:code).scoped_to(:enterprise_id) }
```
Comment (mkllnk) [inline spec/models/voucher_spec.rb:16]: I was surprised that you defined the subject because most shoulda matchers don't need that. It's only the uniqueness which needs that. So strictly speaking, the presence test doesn't need the subject and we don't need to build that enterprise. But since you are only building it in memory and these specs execute in less than a second it doesn't matter. Good solution.

File: app/controllers/admin/vouchers_controller.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+module Admin
+  class VouchersController < ResourceController
+
+   def new
+     @enterprise = Enterprise.find_by permalink: params[:enterprise_id]
```
Comment (mkllnk) [inline app/controllers/admin/vouchers_controller.rb]: Rubocop isn't enforcing it but we prefer parenthesis around arguments.

File: app/views/admin/enterprises/form/_vouchers.html.haml [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        %th= t('.purpose')
+        %th= t('.expiry')
+        %th= t('.use_limit')
+        %th= t('.customers')
+        %th= t('.net_value')
+    %tbody
+      - @enterprise.vouchers.each do |voucher|
+        %tr
+          %td= voucher.code
+          %td= voucher.rate
```
Comment (mkllnk) [inline app/views/admin/enterprises/form/_vouchers.html.haml]: There's no `rate` on the voucher yet in this commit.

File: app/views/admin/vouchers/new.html.haml [hunk 4]
```diff
@@ -1,22 +1,23 @@
...
-          = label :voucher, :code, t('.voucher_code')
-        .omega.eight.columns
-          %textarea.fullwidth{ id: 'voucher_code', name: 'voucher_code', rows: 6 }
-      .row
-        .alpha.four.columns
-          = label :voucher, :amount, t('.voucher_amount')
-        .omega.eight.columns
-          $
-          %input{ type: 'text', id: 'voucher_amount', name: 'voucher_amount', value: 10, disabled: true }
+= form_with model: @voucher, url: admin_enterprise_vouchers_path(@enterprise), html: { name: "voucher_form" } do |f|
```
Comment (mkllnk) [inline app/views/admin/vouchers/new.html.haml:1]: Why does it have a name?

File: spec/javascripts/stimulus/tabs_and_panels_controller_test.js [hunk 5]
```diff
@@ -2,16 +2,26 @@
  * @jest-environment jsdom
  */
 
-import { Application } from "stimulus";
-import tabs_and_panels_controller from "../../../app/webpacker/controllers/tabs_and_panels_controller";
+import { Application } from 'stimulus';
+import tabs_and_panels_controller from '../../../app/webpacker/controllers/tabs_and_panels_controller';
```
Comment (mkllnk) [inline spec/javascripts/stimulus/tabs_and_panels_controller_test.js:6]: Why single quotes? I don't think that we have a convention here. Maybe we need one... I personally prefer double quotes. :wink:

File: config/routes/admin.rb [hunk 6]
```diff
@@ -41,8 +41,9 @@
 
       resources :tag_rules, only: [:destroy]
 
-      # TODO do we need to remove more routes
-      resources :vouchers, only: [:new, :create]
+      constraints FeatureToggleConstraint.new(:vouchers) do
+        resources :vouchers, only: [:new, :create]
+      end
```
Comment (mkllnk) [inline config/routes/admin.rb:46]: I don't think that we need to restrict routes. It's not that people are not allowed to use it. We just don't want to show it yet because it's experimental.

**Review by mkllnk — APPROVED** [review]
Great work!

I added a few comments but it's all unimportant detail. Well done.

File: spec/models/voucher_spec.rb (hunk 1 above)
Comment (rioug) [reply spec/models/voucher_spec.rb:16]: How would you do it then ? Would you just build the subject for the uniqueness `it` block ?

File: app/views/admin/vouchers/new.html.haml (hunk 4 above)
Comment (rioug) [reply app/views/admin/vouchers/new.html.haml:1]: No particular reason, I must copy pasted the `form_with` from somewhere else.

File: spec/javascripts/stimulus/tabs_and_panels_controller_test.js (hunk 5 above)
Comment (rioug) [reply spec/javascripts/stimulus/tabs_and_panels_controller_test.js:6]: Old habits, I tend to prefer single quotes :smile:  It probably comes from my PHP days.

File: spec/models/voucher_spec.rb (hunk 1 above)
Comment (mkllnk) [reply spec/models/voucher_spec.rb:16]: Yes, I would have defined it in there:

```rb
it "validates code uniqueness per enterprise" do
  subject = Voucher.new(code: 'new_code', enterprise: enterprise)
  expect(subject).to validate_uniqueness_of(:code).scoped_to(:enterprise_id)
end
```

But I don't think it matters here. Your solution is more elegant in general and looks better. Maybe it's good to always supply valid values to the matchers. It provides an example and doesn't rely on Shoulda to generate random valid values. :shrug: 

I think that my example above would only be better if it improved the execution time significantly.

File: app/views/admin/enterprises/form/_vouchers.html.haml (hunk 3 above)
Comment (rioug) [reply app/views/admin/enterprises/form/_vouchers.html.haml]: True I am not sure what I did there. Anyway it has been replaced by `voucher.display_value` in this commit https://github.com/openfoodfoundation/openfoodnetwork/pull/10523/commits/e669bb06dc728e6ad9beba7bf9db23a35b4fe6c9

File: spec/javascripts/stimulus/tabs_and_panels_controller_test.js [hunk 7]
```diff
@@ -26,23 +36,75 @@ describe("EnterprisePanelController", () => {
...
+        ka.click();
+        expect(ka.classList.contains('selected')).toBe(true);
+      });
+    })
+
+    describe('when anchor is specified in the url', () => {
+      const { location } = window;
+      const mockLocationToString = (panel) => {
+        // Mocking window.location.toString() 
+        const url = `http://localhost:3000/admin/enterprises/great-shop/edit#!#${panel}`
```
Comment (dacook) [inline spec/javascripts/stimulus/tabs_and_panels_controller_test.js:68]: Note that this might be affected by [this PR](https://github.com/openfoodfoundation/openfoodnetwork/pull/9729): 
> Remove #!# from URLs as they aren't valid.

**Review by dacook — CHANGES_REQUESTED** [review]
Nice, great to have step one!

`±` I think I might have found a small issue in the JS, can you please take a look?

Also a note that I found it hard to review commit-by-commit, because there was a lot of work-in-progress that kept changing, and I had to look to future commits to see if an issue got fixed or not. So I tried reviewing the entire 'files changed' view, but without the context of the commit messages this was also hard.

It's not easy to have perfect commit history, that's for sure. You might be familiar with this already, but in case it's helpful:
* I always make sure to `git diff` before staging, to review if anything should be excluded.
* A tip (that I think Maikel taught me) is to use `git add --interactive`, which you can use to add just parts of files for a commit.
* I often find that helpful along with `git commit --amend` to add little fixups to the latest commit.

File: app/webpacker/controllers/tabs_and_panels_controller.js [hunk 8]
```diff
@@ -12,13 +12,34 @@ export default class extends Controller {
...
+  updateActivePanel(panel_id) {
     const newActivePanel = this.panelTargets.find(
-      (panel) => panel.id == `${event.currentTarget.id}_panel`
+      (panel) => panel.id == panel_id
     );
 
+    if (newActivePanel === undefined) {
+      // No panel found
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (rioug) [reply app/webpacker/controllers/tabs_and_panels_controller.js:41]: `find` does return `undefined` if nothing found see : https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find
But you are right, this scenario isn't covered by the specs, the one on line 55 is though.

File: app/controllers/admin/vouchers_controller.rb [hunk 9]
```diff
@@ -2,10 +2,35 @@
...
+      voucher_params = permitted_resource_params.merge(enterprise: @enterprise)
+      @voucher = Voucher.create(voucher_params)
+
+      if @voucher.save
+        redirect_to(
+          "#{edit_admin_enterprise_path(@enterprise)}#vouchers_panel",
+          flash: { success: flash_message_for(@voucher, :successfully_created) }
+        )
+      else
+        flash[:error] = @voucher.errors.full_messages.to_sentence
```
[the remark this answers is not in the document]
Comment (rioug) [reply app/controllers/admin/vouchers_controller.rb:21]: Indeed you'll be testing a Rails functionality.

File: spec/javascripts/stimulus/tabs_and_panels_controller_test.js (hunk 7 above)
Comment (rioug) [reply spec/javascripts/stimulus/tabs_and_panels_controller_test.js:68]: This isn't a problem. 
The URL would be come something like `http://localhost:3000/admin/enterprises/great-shop/edit#${panel}`
and the following code will picks up the string after the last `#` :

https://github.com/rioug/openfoodnetwork/blob/b0462983d747cbd5d86839df047262f87108983f/app/webpacker/controllers/tabs_and_panels_controller.js#L17-L18

File: spec/system/admin/vouchers_spec.rb [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+    # Then I see a list of vouchers
+    expect(page).to have_content voucher_code
+    expect(page).to have_content "10"
+  end
+
+  it 'creates a voucher' do
+    # Given an enterprise
+    # When I go to the new voucher page
+    login_as_admin_and_visit new_admin_enterprise_voucher_path(enterprise)
```
[the remark this answers is not in the document]
Comment (rioug) [reply spec/system/admin/vouchers_spec.rb]: Good find :+1:

File: spec/models/voucher_spec.rb [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+require 'spec_helper'
+
+describe Voucher do
+  describe 'associations' do
+    it { is_expected.to belong_to(:enterprise) }
+  end
+
+  describe 'validations' do
+    subject { Voucher.new(code: 'new_code', enterprise: enterprise) }
```
[the remark this answers is not in the document]
Comment (rioug) [reply spec/models/voucher_spec.rb:11]: Good point :+1: , added a comment on the original issue.

File: app/webpacker/controllers/tabs_and_panels_controller.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (dacook) [reply app/webpacker/controllers/tabs_and_panels_controller.js:41]: Oh, thanks for pointing that out. I think I need to brush up on my javascript!

**Review by dacook — APPROVED** [review]
👍 Great!

**Comment by filipefurtad0** [comment]
- Visited the Enterprise tab and clicked on "Settings"
- Checked that 'vouchers' appears in the left menu column, and click on it
    --> It shows a page saying "No vouchers yet" and have "Add New" button :tada: 

image

- Clicking on "Add New button"
    --> It should load the New Voucher page

This, however, this triggers a `401 Unauthorized` error:

Below we see - left, the feature toggle with the feature fully enabled; right the admin session, after clicking "Add new": 
image

So, basically what happens is:

`/admin/enterprises/<hub-name>/vouchers/new`  -> redirects to -> `/unauthorized`

Should we merge and address it behind the feature toggle - or rather move it back to in dev? What's the best way to further work on the feature @rioug ? Thanks for your work and feedback here.

File: spec/javascripts/stimulus/tabs_and_panels_controller_test.js (hunk 7 above)
Comment (rioug) [reply spec/javascripts/stimulus/tabs_and_panels_controller_test.js]: @dacook turned out you were right, the PR did break my original solution :disappointed:

**Comment by rioug** [comment]
@filipefurtad0 I actually forgot to add permission for enterprise user. It's pretty much ready to test, but I'll send it back to review just to make sure permission are right. 
Thanks !

**Comment by mkllnk** [comment]
* Good to synch with: #10600

I designed that other PR so that there won't be a conflict but it would be good to include this new feature toggle in the PR that is merged last (this or the feature toggle PR).

File: db/schema.rb [hunk 12]
```diff
@@ -1198,6 +1198,16 @@
     t.index ["user_id"], name: "index_webhook_endpoints_on_user_id"
   end
 
+  create_table "vouchers", force: :cascade do |t|
+    t.string "code", limit: 255, null: false
+    t.datetime "expiry_date"
```
Comment (mkllnk) [inline db/schema.rb:1203]: It's not used yet but I imagine an index on expiry_date very useful, too.

**Review by mkllnk — APPROVED** [review]
Looking good to me.

**Comment by mkllnk** [comment]
@rioug I merged the feature toggle descriptions. So this branch now needs updating to add the feature toggle to the official list in the code.

**Comment by rioug** [comment]
Description added

**Review by mkllnk — APPROVED** [review]

**Review by dacook — APPROVED** [review]
👍 The latest changes look good to me, good to see the feature spec updated.

**Comment by filipefurtad0** [comment]
Hey @rioug ,

I've repeated the tests you've suggested:

- after visiting the Enterprise tab and click on "Settings" for a shop
One can verify that the 'vouchers' section appears in the left menu column

image

One note though: this is appearing for all profiles having access to backoffice, i.e.:
  - producer profile
  - shop profile
  - hub profile

- clicking on it
  - displays "No vouchers yet" and have an "Add New" button
  - it displays the dashboard left-menu

image

- Clicking on "Add New button"
  - loads the New Voucher page
  - it hides the dashboard left-menu
  - I think this page is also missing a header-bar or title (compare with the previous pics, the gray bar where `Settings: Kimchi Hub` is stated)

image
 
- Creating a voucher after adding a code and clicking "Save"
    --> It should redirect you to the vouchers page on the enterprise edit page
    --> You should now see a list of vouchers with the one you just created

image

- Entering a voucher without entering any code
    --> Displays an error message saying "Code can't be blank"

image

**Summary**

I think we're good to merge :+1: 

Some things to note and double check:
- I think producer only profiles don't need to have access to this feature
- a header is missing on the create voucher page

Merging!
